tests/mockserver.py

```diff
@@ -190,17 +197,37 @@ def render(self, request):
 
 class MockServer():
 
+    HTTP = 'mock_http'
+    HTTPS = 'mock_https'
+
+    @classmethod
+    def from_mock(cls, path, is_secure=False):
+        key = cls.HTTP
+        if is_secure:
+            key = cls.HTTPS
+        return os.environ[key] + path
+
     def __enter__(self):
         from scrapy.utils.test import get_testenv
-        self.proc = Popen([sys.executable, '-u', '-m', 'tests.mockserver'],
+
+        http_port = str(get_ephemeral_port())
+        https_port = str(get_ephemeral_port())
+
+        self.proc = Popen([sys.executable, '-u', '-m', 'tests.mockserver', http_port, https_port],
                           stdout=PIPE, env=get_testenv())
         self.proc.stdout.readline()
+
+        self._oldenv = os.environ.copy()
+        os.environ[self.HTTP] = 'http://localhost:%s' % (http_port, )
+        os.environ[self.HTTPS] = 'https://localhost:%s' % (https_port, )
```

Comment:
I think one of the goals should be to allow running Scrapy tests in parallel, including parallel in the same process. For this to work ports shouldn't be global - they should be attributes of the MockServer instance. I know it can be a larger change :(

Reply:
Hm, but using a single MockServer can be fine for Scrapy purposes, so maybe having host/port global is fine.

Reply:
To make port as variable of object you need to pass it to spider classes. And there is no clean way to do this. `scrapy.crawler.CrawlerRunner.create_crawler` does not accept arguments. There are lot of such places.